R/module_nested_tabs.R

```diff
@@ -84,235 +84,238 @@ ui_nested_tabs.teal_modules <- function(id, modules, datasets, depth = 0L, is_mo
-    # We want to recalculate only visible modules
-    # - trigger the data when the tab is selected
-    # - trigger module to be called when the tab is selected for the first time
-    trigger_data <- reactiveVal(1L)
-    trigger_module <- reactiveVal(NULL)
-    output$data_reactive <- renderUI({
-      lapply(datasets$datanames(), function(x) {
-        datasets$get_data(x, filtered = TRUE)
-      })
-      isolate(trigger_data(trigger_data() + 1))
-      isolate(trigger_module(TRUE))
-
-      NULL
+    active_datanames <- reactive({
+      req(inherits(data_rv(), "teal_data"))
+      datanames <- if (is.null(modules$datanames) || identical(modules$datanames, "all")) {
+        teal_data_datanames(data_rv())
+      } else {
+        # Remove datanames that are not **YET** in the data (may be added with teal_data_module transforms)
+        # Check is deferred when module is called
+        intersect(
+          include_parent_datanames(
+            modules$datanames,
+            teal.data::join_keys(data_rv())
+          ),
+          teal_data_datanames(data_rv())
+        )
+      }
     })
-
-    # collect arguments to run teal_module
-    args <- c(list(id = "module"), modules$server_args)
-    if (is_arg_used(modules$server, "reporter")) {
-      args <- c(args, list(reporter = reporter))
-    }
-
-    if (is_arg_used(modules$server, "datasets")) {
-      args <- c(args, datasets = datasets)
+    if (is.null(datasets)) {
+      datasets <- eventReactive(data_rv(), {
+        req(inherits(data_rv(), "teal_data"))
+        logger::log_trace("srv_teal_module@1 initializing module-specific FilteredData")
+        teal_data_to_filtered_data(data_rv(), datanames = active_datanames(), filter = slices_global())
+      })
     }
 
-    if (is_arg_used(modules$server, "data")) {
-      data <- eventReactive(trigger_data(), .datasets_to_data(modules, datasets))
-      args <- c(args, data = list(data))
-    }
+    # manage module filters on the module level
+    # important:
+    #   filter_manager_module_srv needs to be called before filter_panel_srv
+    #   Because available_teal_slices is used in FilteredData$srv_available_slices (via srv_filter_panel)
+    #   and if it is not set, then it won't be available in the srv_filter_panel
+    srv_module_filter_manager(modules$label, module_fd = datasets, slices_global = slices_global)
+    filtered_teal_data <- srv_filter_panel(
+      "filter_panel",
+      datasets = datasets,
+      active_datanames = active_datanames,
+      data_rv = data_rv,
+      is_active = is_active
+    )
 
-    if (is_arg_used(modules$server, "filter_panel_api")) {
-      filter_panel_api <- teal.slice::FilterPanelAPI$new(datasets)
-      args <- c(args, filter_panel_api = filter_panel_api)
-    }
+    transformed_teal_data <- srv_teal_data_modules(
+      "data_transform",
+      data = filtered_teal_data,
+      transformers = modules$transformers,
+      modules = modules
+    )
```

Comment:
Question: Do we want transformers to have an assertion `checkmate::assert_class(isolate(data()), "teal_data")` on transformers just as teal modules?

Currently, it's passing a shiny error, so it's not possible.

Let's say:

```r
transformer_with_assertion <- teal_data_module(
  label = "head",
  ui = function(id) tags$span("Dummy"),
  server = function(id, data) {
    checkmate::assert_class(data, "reactive")
    checkmate::assert_class(isolate(data()), "teal_data")
    moduleServer(id, function(input, output, session) data)
  }
)
```

<details>

<summary>Code to reproduce error</summary>

```r

options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("../teal")

data <- teal_data_module(
  ui = function(id) actionButton(NS(id, "button"), "Click me"),
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$button, {
        data <- teal_data() |>
          within({
            utils::data(miniACC, package = "MultiAssayExperiment", envir = environment())
            MAE <- miniACC
          })
        datanames(data) <- "MAE"
        data
      })
    })
  }
)

transformer_with_assertion <- teal_data_module(
  label = "head",
  ui = function(id) tags$span("Dummy"),
  server = function(id, data) {
    checkmate::assert_class(data, "reactive")
    checkmate::assert_class(isolate(data()), "teal_data")
    moduleServer(id, function(input, output, session) data)
  }
)

init(
  data = data,
  modules = example_module(transformers = list(transformer_with_assertion))
) |>
  shiny::runApp()
```

</details>

Reply:
I don't think it is needed. In the initial `srv_teal_data_module` we "fallback" to NULL, which stops `teal` from further evaluation. Once this initial `srv_teal_data_module` is validated there is no way in `teal` to break due to `.fallback_on_failure`

Reply:
On the initial `datasets` reactive has a `req` that propagates it. Fallback to NULL was a comment after yesterday bug finding mission and removed on [7f99484256ae74](https://github.com/insightsengineering/teal/pull/1253/commits/7f99484256ae74554421f8cbcf086bac86a8b19f#diff-f7278c16079bf8eeee343aba99b73c110c99d6cc705b7bcdbcc0df3c8bde35abL258-L269)

**I do agree**, I raised this question because developers might apply that assertion pattern here out of habit (I know I did in some of my testing applications)